### PR TITLE
Move many udf implementations from `invoke` to `invoke_batch`

### PR DESCRIPTION
**Pull request by joseph-isaacs**
## Which issue does this PR close?

As part of https://github.com/apache/datafusion/issues/13238 invoke was replaced with invoke_batch.
- Part of https://github.com/apache/datafusion/issues/13516
- Closes https://github.com/apache/datafusion/pull/13507
## Rationale for this change

This will allow the removal of invoke. And maybe if we are quick we can rename `invoke_with_args` back to `invoke`. 

## What changes are included in this PR?

This PR moves function definitions over to use invoke_batch, so invoke can be removed.

### CODE REVIEW

File: docs/source/user-guide/sql/scalar_functions.md [hunk 1]
```diff
@@ -1954,32 +1954,6 @@ The following intervals are supported:
 - years
 - century
 
-#### Example
```
Comment (joseph-isaacs) [inline docs/source/user-guide/sql/scalar_functions.md:1957]: ./dev/update_function_docs.sh did this?

File: datafusion/expr/src/udf.rs [hunk 2]
```diff
@@ -546,7 +546,7 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// to arrays, which will likely be simpler code, but be slower.
     fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
         #[allow(deprecated)]
```
Comment (alamb) [inline datafusion/expr/src/udf.rs:548]: we cam probably remove this #allow as well

File: datafusion/functions/benches/ltrim.rs [hunk 3]
```diff
@@ -141,8 +141,8 @@ fn run_with_string_type<M: Measurement>(
         ),
         |b| {
             b.iter(|| {
-                #[allow(deprecated)] // TODO use invoke_batch
-                black_box(ltrim.invoke(&args))
+                #[allow(deprecated)] // TODO use invoke_with_args
```
Comment (alamb) [inline datafusion/functions/benches/ltrim.rs:144]: likewise here I think we can remove the #allow

File: datafusion/functions/benches/make_date.rs [hunk 4]
```diff
@@ -91,13 +101,15 @@ fn criterion_benchmark(c: &mut Criterion) {
         let mut rng = rand::thread_rng();
         let year = ColumnarValue::Scalar(ScalarValue::Int32(Some(2025)));
         let month = ColumnarValue::Scalar(ScalarValue::Int32(Some(11)));
-        let days = ColumnarValue::Array(Arc::new(days(&mut rng)) as ArrayRef);
+        let day_arr = Arc::new(days(&mut rng));
+        let batch_len = day_arr.len();
+        let days = ColumnarValue::Array(day_arr);
 
         b.iter(|| {
             #[allow(deprecated)] // TODO use invoke_batch
```
Comment (alamb) [inline datafusion/functions/benches/make_date.rs:109]: comments could be updated (todo removed)

File: datafusion/functions/src/datetime/date_bin.rs [hunk 5]
```diff
@@ -181,30 +185,6 @@ Calculates time intervals and returns the start of the interval nearest to the s
 For example, if you "bin" or "window" data into 15 minute intervals, an input timestamp of `2023-01-01T18:18:18Z` will be updated to the start time of the 15 minute bin it is in: `2023-01-01T18:15:00Z`.
 "#)
             .with_syntax_example("date_bin(interval, expression, origin-timestamp)")
-            .with_sql_example(r#"```sql
```
Comment (alamb) [inline datafusion/functions/src/datetime/date_bin.rs:184]: why was this removed?

File: datafusion/functions/src/utils.rs [hunk 6]
```diff
@@ -149,7 +149,8 @@ pub mod test {
                     let return_type = return_type.unwrap();
                     assert_eq!(return_type, $EXPECTED_DATA_TYPE);
 
-                    let result = func.invoke_with_args(datafusion_expr::ScalarFunctionArgs{args: $ARGS, number_rows: cardinality, return_type: &return_type});
+                    #[allow(deprecated)]
```
Comment (alamb) [inline datafusion/functions/src/utils.rs:152]: It seems like it would be good to keep calling using `invoke_with_args`

File: datafusion/functions/src/utils.rs [hunk 7]
```diff
@@ -170,7 +171,8 @@ pub mod test {
                     }
                     else {
                         // invoke is expected error - cannot use .expect_err() due to Debug not being implemented
-                        match func.invoke_with_args(datafusion_expr::ScalarFunctionArgs{args: $ARGS, number_rows: cardinality, return_type: &return_type.unwrap()}) {
+                        #[allow(deprecated)]
```
Comment (alamb) [inline datafusion/functions/src/utils.rs:174]: likewise here -- invoke_with_args

File: docs/source/user-guide/sql/scalar_functions.md (hunk 1 above)
Comment (alamb) [reply docs/source/user-guide/sql/scalar_functions.md:1957]: I think it is because you removed the docs (perhaps accidentally) above -- i left a comment

**Review by alamb — COMMENTED** [review]
Thanks @joseph-isaacs  -- this looks like a great step forward

I had some suggestions / additional cleanups

THe only thing I think that needs to be fixed before merging this is reverting the loss of the documentation for date_bin

File: datafusion/functions/src/utils.rs (hunk 7 above)
Comment (joseph-isaacs) [reply datafusion/functions/src/utils.rs:174]: A lot of tests use this macro and pass in &[], so it quite a big change.

File: datafusion/functions/benches/ltrim.rs (hunk 3 above)
Comment (joseph-isaacs) [reply datafusion/functions/benches/ltrim.rs:144]: Comment below

File: datafusion/functions/src/datetime/date_bin.rs (hunk 5 above)
Comment (joseph-isaacs) [reply datafusion/functions/src/datetime/date_bin.rs]: Oops, bad merge

File: datafusion/expr/src/udf.rs [hunk 8]
```diff
@@ -213,7 +213,6 @@ impl ScalarUDF {
         self.inner.is_nullable(args, schema)
     }
 
-    #[deprecated(since = "43.0.0", note = "Use `invoke_with_args` instead")]
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/udf.rs:216]: Yes I think it is fine / correct

**Review by alamb — COMMENTED** [review]
Thanks @joseph-isaacs  -- this is looking close to me, though it looks like a few bugs got introduced. I flagged some of them.

I can try to help tomorrow, but I am out of time today

**Comment by alamb** [comment]
Here is what I suggest we do with this PR:
1. Remove the change for https://github.com/apache/datafusion/pull/13507 (let's do that in a separate PR)
2. Fix up the regressions introduced (where batch_size --> 1)

Then merge this one in. 

If you have more time @joseph-isaacs  it would be even better to change this PR so all the functions used `ScalarFunctionArgs` but I realize that is asking quite a bit

**Comment by joseph-isaacs** [comment]
I would have liked to move over to invoke_with_args, however its quite a bit more of an undertaking. Possibly in the future I can try

File: datafusion/expr/src/udf.rs [hunk 9]
```diff
@@ -330,7 +328,7 @@ where
 
 pub struct ScalarFunctionArgs<'a> {
     // The evaluated arguments to the function
-    pub args: &'a [ColumnarValue],
+    pub args: Vec<ColumnarValue>,
```
[the remark this answers is not in the document]
Comment (joseph-isaacs) [reply datafusion/expr/src/udf.rs:331]: I am happy to remove this, but it will unwind a lot of tests changes needed to allow this change.

File: datafusion/functions/src/datetime/date_bin.rs [hunk 10]
```diff
@@ -544,7 +548,7 @@ mod tests {
                 ColumnarValue::Array(timestamps),
                 ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(1), None)),
             ],
-            batch_size,
+            1,
```
[the remark this answers is not in the document]
Comment (joseph-isaacs) [reply datafusion/functions/src/datetime/date_bin.rs]: I had a look for all of these, but its easy to miss one, all the others that pass an literal value take only scalars.

**Comment by alamb** [comment]
> I would have liked to move over to invoke_with_args, however its quite a bit more of an undertaking. Possibly in the future I can try

Makes sense

**Comment by joseph-isaacs** [comment]
I think there was a transient failure

**Comment by alamb** [comment]
> I think there was a transient failure

I retriggered the tests

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @joseph-isaacs

**Comment by jayzhan211** [comment]
Thanks @joseph-isaacs @alamb

**Comment by alamb** [comment]
🙏 thank you -- we are making progresss
